raft/src/main/java/org/apache/kafka/raft/CandidateState.java

```diff
@@ -241,12 +236,14 @@ public int epoch() {
     }
 
     @Override
-    public boolean canGrantVote(int candidateId, boolean isLogUpToDate) {
+    public Optional<String> validateGrantVote(int candidateId, boolean isLogUpToDate) {
         // Still reject vote request even candidateId = localId, Although the candidate votes for
         // itself, this vote is implicit and not "granted".
-        log.debug("Rejecting vote request from candidate {} since we are already candidate in epoch {}",
-            candidateId, epoch);
-        return false;
+        return Optional.of(String.format(
+            "Rejecting vote request from candidate %s since we are already candidate in epoch %s",
+            candidateId,
+            epoch
+        ));
```

Comment:
One tricky aspect is that we now generate the string even if logging is disabled. Let me think about that a bit more.

Reply:
This is tricky, fortunately, vote request is not a frequent RPC.